### DISCUSSION TITLE
Support for different vercel.json/now.json for each deployment type (production, preview, development)

### DISCUSSION THREAD

**Discussion by gniquil** [vercel/vercel#4677] [Ideas]
Currently we can specify a `now.json` or `vercel.json` at the root of the project to support "multi-zone" setup via proxy. The problem with this is we can only have 1 version of this file. This forces us to proxy to the same destination regardless whether this is a production deploy or a preview (or staging type) deploy.

Based on how Vercel's environment variables work, I'd like to propose to have something similar:

- depending on the type of deploy, we would first look a `now.[deploy-type].json` e.g. `now.production.json` for production and `now.preview.json` for preview, and finally `now.development.json` when running `now dev`
- finally, if `now.json` or `vercel.json` also exists, we would merge the contents similar to how environments work.

What do you guys think?

Frank

**Comment by mcsdevv** [comment]
Hello Frank,

Can you tell me what it is that's missing from [Vercel Environment Variables](https://vercel.com/docs/v2/build-step#environment-variables) that would stop you achieving the desired behavior?

We recently updated them to provide support for multiple environments.

**Reply by gniquil** [reply]
Hi,

Thanks for the question. The problem with environment variables is they can't be used in 'vercel.json' files. More specially I need to proxy traffic to different destinations depending on different environments to integrate existing services (production vs staging).

Frank

**Reply by styfle** [reply]
If you're using Next.js, you can install the latest canary `npm i -g next@canary` and utilize dynamic rewrites in a `next.confg.js` file:

```js
module.exports = {
  experimental: {
    rewrites() {
      return [
        { source: "/upstream/:match*", destination: `${process.env.UPSTREAM_URL}/api/:match*` }
      ]
    },
  },
};
```

**Reply by gniquil** [reply]
WOW! I guess this is the thing I've been waiting for. Are you guys gonna make an official announcement soon? Is there a place where I can look at more detailed docs?

**Reply by gniquil** [reply]
Hi,

I guess I have to tame my enthusiasm a bit here. I just gave this a try. It behaved a bit differently than what I expected. 

With rewrite above, some of the proxied sites get redirected back to the proxied server. I then ran it using vercel with vercel.json file and that worked fine. 

Should you guys just implement all the features supported using "routes" config in `vercel.json`?

**Reply by styfle** [reply]
This will be announced and documented once it reached stable.

The RFC is here with some details: https://github.com/vercel/next.js/discussions/9081

**Reply by ijjk** [reply]
@gniquil can you provide a reproduction for where this is behaving differently when used in `next.config.js` vs `vercel.json` we would be happy to take a look and address this

**Reply by jakst** [reply]
I keep seeing references to this functionality coming up in next.js when the questions is asked about support in Vercel. This is of course great, but not everybody uses Vercel to deploy Next apps, and sometimes it feels invasive to mix up application behaviour with platform behaviour.

Routing and proxying should be more ambitious on the Vercel level than on the next.js level in my opinion.

**Reply by jozsefsallai** [reply]
Apologies for necroposting, however, I would like to ask if this is still not possible in Vercel (not Next.js).

The use case here would be that we have a production domain (e.g. acme.ex) which is allowed by CORS to access endpoints on api.acme.ex. However, we do not allow any other domains to access our API. For this purpose, we want to proxy requests coming to `vercel-preview-deployment-url.vercel.app/proxy` to the API so that we can use the app in the preview environment. We can use rewrites to achieve this of course, but we don't want to have the proxy URL in the production environment.

Is there a sane way to accomplish this at a platform-level? I am aware that I can create multiple configurations (i.e. `vercel.json` and `vercel.prod.json`) and use `vercel -A` to point to the one I want to use, but is there a way to specify which one to use when deploying from GitHub for example?
